File: 133-clone-graph/133-clone-graph.rb
# Definition for a Node.
# class Node
#     attr_accessor :val, :neighbors
#     def initialize(val = 0, neighbors = nil)
#		  @val = val
#		  neighbors = [] if neighbors.nil?
#         @neighbors = neighbors
#     end
# end

# @param {Node} node
# @return {Node}
def cloneGraph(node)
    return node if !node
    dfs node, map = [nil] * 100
    map[node.val]
end

def dfs(node, map)
    return map[node.val] if map[node.val]
    
    copy = Node.new node.val
    map[node.val] = copy
    node.neighbors.each do |neighbor|
        copy.neighbors << dfs(neighbor, map)
    end
    copy
end

File: medium/133-clone-graph/133-clone-graph.js
/**
 * // Definition for a Node.
 * function Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {Node} node
 * @return {Node}
 */

// use a dfs/bfs and hashmap

// for each node, add its copy as 
// value in hashmap (created)
// check if already cloned
// if not, make a copy and push
// copies of neighbor into clone.neighbor
const cloneGraph = function(node) {
    if (!node) return null;
    const map = new Map();
    
    (function dfs(node) {
        if (map.has(node)) return map.get(node);
        
        const copy = new Node(node.val);
        map.set(node, copy);
        node.neighbors.forEach(neighbor => {
            copy.neighbors.push(dfs(neighbor));
        })
        return copy;
    }(node))
    
    return map.get(node)
};